epee: cleanup string_tools

0xFFFC0000 opened the pull request:
1. Use `boost::filesystem` for already available operations.
2. Use `boost::string` for already available operations.
3. ~~Fix the inconsistent indentation.~~

vtnerd commented on contrib/epee/include/string_tools.h:
```diff
@@ -47,12 +48,12 @@ namespace epee
 {
 namespace string_tools
 {
-  //----------------------------------------------------------------------------
```
We have historically refrained from changing whitespace comments (or in this case decorator) such as this.

0xFFFC0000 replied:
Returned all the non-code changes to previous state.

vtnerd commented on contrib/epee/src/string_tools.cpp:
```diff
@@ -112,81 +116,70 @@ namespace string_tools
-	
+  
+  bool compare_no_case(const std::string& str1, const std::string& str2)
+  {
```
Again more whitespace changes. These lines will now attribute you, even though nothing much as been done.

Summoning @moneromooo-monero who used to push back on these.

0xFFFC0000 replied:
Returned all the non-code changes to the previous state. Thanks.

vtnerd commented on contrib/epee/src/string_tools.cpp:
```diff
@@ -112,81 +116,70 @@ namespace string_tools
-			return false;
 
-	}
+    get_current_module_name() = path_to_process.filename().string();
```
Now this is a nice cleanup.

vtnerd commented on contrib/epee/src/string_tools.cpp:
```diff
@@ -112,81 +116,70 @@ namespace string_tools
+    buf.resize(MAX_PATH);
+    do
+    {
+        unsigned int len = GetModuleFileNameW(NULL, &buf[0], static_cast< unsigned int >(buf.size()));
```
This also returns `0` on error, I think it just loops until buff exceeds `65536` in that case. Perhaps an explicit check? We could `throw` I guess, although that changes the functionality a bit.

vtnerd reviewed (commented):
The only hold up are the whitespace changes, I roped in @moneromooo-monero hopefully who constantly pushed back against such changes.

steverusso commented on contrib/epee/src/string_tools.cpp:
```diff
@@ -141,52 +144,55 @@ namespace string_tools
+  void trim_left(std::string& str)
+  {
+    boost::trim_left(str);
+    return;
```
Should this redundant `return;` statement be removed? I believe the `void` function will return here anyway.

(If so, this is also the case in the `trim_right` function a few lines below.)

moneromooo-monero reviewed (commented):
I changed my mind. It's not that large, and I see at least one improvement so it's not just "I like it better that way".

0xFFFC0000 commented:
Changelog: 
1. Cleaned unnecessary windows function change.
2. Rebased to upstream/master. 
3. Changed the PR name to reflect our rules for PR names.

tobtoht commented:
This patch caused some breakage on Windows.

If the path to the executable contains unicode characters (e.g. `é`) the process will terminate with:

```
    terminate called after throwing an instance of 'boost::system::system_error'
      what():  boost::filesystem::path codecvt to wstring: error [codecvt:2]
```

On this line: https://github.com/monero-project/monero/blame/3b01c490953fe92f3c6628fa31d280a4f0490d28/contrib/epee/src/string_tools.cpp#L159

aleflm commented:
> This patch caused some breakage on Windows.
> 
> If the path to the executable contains unicode characters (e.g. `é`) the process will terminate with:
> 
> ```
>     terminate called after throwing an instance of 'boost::system::system_error'
>       what():  boost::filesystem::path codecvt to wstring: error [codecvt:2]
> ```
> 
> On this line: https://github.com/monero-project/monero/blame/3b01c490953fe92f3c6628fa31d280a4f0490d28/contrib/epee/src/string_tools.cpp#L159

path_to_process = get_current_module_path(); <- This is the likely culprit. You are assigning a std::string (returned by get_current_module_path(), likely ANSI-encoded on Windows if containing non-ASCII chars like 'é') directly to a boost::filesystem::path. Boost.Filesystem's assignment operator/ constructor taking a std::string assumes the string is UTF-8 encoded by default. If the path contains 'é' (byte OxE9 in Windows-1252), Boost tries to interpret OxE9 as UTF-8, finds it invalid, and throws the codecvt exception during this assignment or when trying to use the path later. 

Give me a day or two. I will fix this.

0xFFFC0000 commented:
> This patch caused some breakage on Windows.
> 
> If the path to the executable contains unicode characters (e.g. `é`) the process will terminate with:
> 
> ```
>     terminate called after throwing an instance of 'boost::system::system_error'
>       what():  boost::filesystem::path codecvt to wstring: error [codecvt:2]
> ```
> 
> On this line: https://github.com/monero-project/monero/blame/3b01c490953fe92f3c6628fa31d280a4f0490d28/contrib/epee/src/string_tools.cpp#L159

path_to_process = get_current_module_path(); <- This is the likely culprit. You are assigning a std::string (returned by get_current_module_path(), likely ANSI-encoded on Windows if containing non-ASCII chars like 'é') directly to a boost::filesystem::path. Boost.Filesystem's assignment operator/ constructor taking a std::string assumes the string is UTF-8 encoded by default. If the path contains 'é' (byte OxE9 in Windows-1252), Boost tries to interpret OxE9 as UTF-8, finds it invalid, and throws the codecvt exception during this assignment or when trying to use the path later. 

Give me a day or two. I will fix this.

0xFFFC0000 commented:
> This patch caused some breakage on Windows.
> 
> If the path to the executable contains unicode characters (e.g. `é`) the process will terminate with:
> 
> ```
>     terminate called after throwing an instance of 'boost::system::system_error'
>       what():  boost::filesystem::path codecvt to wstring: error [codecvt:2]
> ```
> 
> On this line: https://github.com/monero-project/monero/blame/3b01c490953fe92f3c6628fa31d280a4f0490d28/contrib/epee/src/string_tools.cpp#L159

I am unable to reproduce it. I built guix binaries (`master`), and didn't have any issue running `monerod.exe` binary from it from the folder named `mo é É` or `mon Éro` or `É`.  All of them worked without any issue.
